File: packages/totem-sdk/packages/provider-bond/src/scoring.ts
import type {
  ProviderScore,
  ProviderRecommendation,
  ComputeProviderScoreParams,
  ProbeResult,
  IncidentRecord,
} from './types.js';
import { DEFAULT_PROVIDER_SCORING_WEIGHTS, RECOMMENDATION_THRESHOLDS, MAX_HEARTBEAT_AGE_MS_DEFAULT } from './constants.js';

export function computeProviderScore(params: ComputeProviderScoreParams): ProviderScore {
  const { provider, bondProofs, probes, incidents, now, weights } = params;
  const w = weights ?? DEFAULT_PROVIDER_SCORING_WEIGHTS;
  const ts = now ?? Date.now();

  const reasons: string[] = [];

  let identityScore = 0;
  if (provider.providerBond.bondOwnerAddress) {
    identityScore = 100;
    reasons.push('Identity address declared');
  } else {
    reasons.push('No identity address declared');
  }

  let bondScore = 0;
  const hasBond = provider.providerBond.bondStack && provider.providerBond.bondStack.length > 0;
  const hasMinimaBond = hasBond && provider.providerBond.bondStack!.some(
    (d) => d.asset === 'MINIMA' && d.purpose === 'hard-collateral' && d.status === 'active'
  );
  const hasProof = bondProofs && bondProofs.length > 0;

  if (hasMinimaBond && hasProof) {
    bondScore = 100;
    reasons.push('Active MINIMA hard-collateral bond with proof');
  } else if (hasMinimaBond) {
    bondScore = 70;
    reasons.push('MINIMA bond declared but no proof attached');
  } else if (hasBond) {
    bondScore = 40;
    reasons.push('Bond declared but not MINIMA hard-collateral');
  } else {
    reasons.push('No bond declared');
  }

  let reliabilityScore = 50;
  const providerProbes = probes ?? [];
  if (providerProbes.length > 0) {
    const recentProbes = providerProbes.filter((p) => ts - p.observedAt < MAX_HEARTBEAT_AGE_MS_DEFAULT);
    if (recentProbes.length > 0) {
      const successRate = recentProbes.filter((p) => p.ok).length / recentProbes.length;
      const avgLatency = recentProbes.reduce((sum, p) => sum + (p.latencyMs ?? 0), 0) / recentProbes.length;

      reliabilityScore = successRate * 100;
      if (avgLatency > 1000) reliabilityScore -= 20;
      else if (avgLatency > 500) reliabilityScore -= 10;

      reasons.push(`Probe success rate: ${(successRate * 100).toFixed(0)}%`);
    } else {
      reliabilityScore = 0;
      reasons.push('No recent probes (offline)');
    }
  } else {
    reasons.push('No probe data');
  }

  let incidentScore = 100;
  const providerIncidents = incidents ?? [];
  if (providerIncidents.length > 0) {
    const openIncidents = providerIncidents.filter((i) => i.status === 'open');
    const criticalIncidents = providerIncidents.filter((i) => i.severity === 'critical');
    const highIncidents = providerIncidents.filter((i) => i.severity === 'high');

    incidentScore = 100 - (openIncidents.length * 10) - (criticalIncidents.length * 20) - (highIncidents.length * 10);
    incidentScore = Math.max(0, incidentScore);

    if (criticalIncidents.length > 0) reasons.push(`${criticalIncidents.length} critical incident(s)`);
    if (highIncidents.length > 0) reasons.push(`${highIncidents.length} high-severity incident(s)`);
    if (openIncidents.length > 0) reasons.push(`${openIncidents.length} open incident(s)`);
  } else {
    reasons.push('No incidents');
  }

  const totalScore = Math.round(
    identityScore * w.identity +
    bondScore * w.bond +
    reliabilityScore * w.reliability +
    incidentScore * w.incidents
  );

  const recommendation = computeProviderRecommendation(totalScore);

  return {
    providerId: provider.providerBond.providerId,
    score: totalScore,
    recommendation,
    bondScore,
    identityScore,
    reliabilityScore,
    incidentScore,
    computedAt: ts,
    reasons,
  };
}

export function computeProviderRecommendation(score: number): ProviderRecommendation {
  if (score >= RECOMMENDATION_THRESHOLDS.recommended) return 'recommended';
  if (score >= RECOMMENDATION_THRESHOLDS.acceptable) return 'acceptable';
  if (score >= RECOMMENDATION_THRESHOLDS.risky) return 'risky';
  if (score >= RECOMMENDATION_THRESHOLDS.avoid) return 'avoid';
  if (score > 0) return 'unbonded';
  return 'offline';
}

File: packages/totem-sdk/packages/provider-bond/rust/src/scoring.rs
use crate::types::*;

const MAX_HEARTBEAT_AGE_MS_DEFAULT: u64 = 120_000;

const SEVERITY_ORDER: [(&str, u32); 4] = [
    ("low", 0),
    ("medium", 1),
    ("high", 2),
    ("critical", 3),
];

fn severity_value(severity: &str) -> u32 {
    SEVERITY_ORDER.iter().find(|(s, _)| *s == severity).map(|(_, v)| *v).unwrap_or(0)
}

pub fn compute_provider_score(
    provider: &ProviderBondManifest,
    probes: &[ProbeResult],
    incidents: &[IncidentRecord],
    now: u64,
    weights: &ScoringWeights,
) -> ProviderScore {
    let mut reasons: Vec<String> = Vec::new();

    let identity_score: u32 = if provider.provider_bond.bond_owner_address.is_some() {
        reasons.push("Identity address declared".to_string());
        100
    } else {
        reasons.push("No identity address declared".to_string());
        0
    };

    let has_bond = provider.provider_bond.bond_stack.as_ref().map_or(false, |s| !s.is_empty());
    let has_minima_bond = provider.provider_bond.bond_stack.as_ref().map_or(false, |s| {
        s.iter().any(|d| d.asset == "MINIMA" && d.purpose == "hard-collateral" && d.status == "active")
    });

    let bond_score: u32 = if has_minima_bond {
        reasons.push("Active MINIMA hard-collateral bond with proof".to_string());
        100
    } else if has_bond {
        reasons.push("Bond declared but not MINIMA hard-collateral".to_string());
        40
    } else {
        reasons.push("No bond declared".to_string());
        0
    };

    let mut reliability_score: u32 = 50;
    if !probes.is_empty() {
        let recent: Vec<&ProbeResult> = probes.iter().filter(|p| now - p.observed_at < MAX_HEARTBEAT_AGE_MS_DEFAULT).collect();
        if !recent.is_empty() {
            let success_rate = recent.iter().filter(|p| p.ok).count() as f64 / recent.len() as f64;
            let avg_latency: f64 = recent.iter().map(|p| p.latency_ms.unwrap_or(0) as f64).sum::<f64>() / recent.len() as f64;

            reliability_score = (success_rate * 100.0) as u32;
            if avg_latency > 1000.0 { reliability_score = reliability_score.saturating_sub(20); }
            else if avg_latency > 500.0 { reliability_score = reliability_score.saturating_sub(10); }

            reasons.push(format!("Probe success rate: {:.0}%", success_rate * 100.0));
        } else {
            reliability_score = 0;
            reasons.push("No recent probes (offline)".to_string());
        }
    } else {
        reasons.push("No probe data".to_string());
    }

    let mut incident_score: u32 = 100;
    if !incidents.is_empty() {
        let open_count = incidents.iter().filter(|i| i.status == "open").count() as u32;
        let critical_count = incidents.iter().filter(|i| i.severity == "critical").count() as u32;
        let high_count = incidents.iter().filter(|i| i.severity == "high").count() as u32;

        incident_score = 100u32.saturating_sub(open_count * 10).saturating_sub(critical_count * 20).saturating_sub(high_count * 10);

        if critical_count > 0 { reasons.push(format!("{} critical incident(s)", critical_count)); }
        if high_count > 0 { reasons.push(format!("{} high-severity incident(s)", high_count)); }
        if open_count > 0 { reasons.push(format!("{} open incident(s)", open_count)); }
    } else {
        reasons.push("No incidents".to_string());
    }

    let total_score = (
        identity_score as f64 * weights.identity +
        bond_score as f64 * weights.bond +
        reliability_score as f64 * weights.reliability +
        incident_score as f64 * weights.incidents
    ).round() as u32;

    let recommendation = compute_recommendation(total_score);

    ProviderScore {
        provider_id: provider.provider_bond.provider_id.clone(),
        score: total_score,
        recommendation,
        bond_score,
        identity_score,
        reliability_score,
        incident_score,
        computed_at: now,
        reasons,
    }
}

pub fn compute_recommendation(score: u32) -> String {
    (if score >= 80 { "recommended" }
    else if score >= 60 { "acceptable" }
    else if score >= 40 { "risky" }
    else if score >= 20 { "avoid" }
    else if score > 0 { "unbonded" }
    else { "offline" }).to_string()
}
